nvflare/app_common/ccwf/client_controller_executor.py

```diff
@@ -0,0 +1,356 @@
+        self.final_result_ack_timeout = final_result_ack_timeout
+        self.persistor_id = persistor_id
+
+        self.persistor = None
+
+        self.current_status = StatusReport()
+        self.last_status_report_time = time.time()  # time of last status report to server
+        self.config = None
+        self.workflow_id = None
+        self.finalize_lock = threading.Lock()
+
+        self.asked_to_stop = False
+        self.status_lock = threading.Lock()
+        self.engine = None
+        self.me = None
+        self.is_starting_client = False
+        self.workflow_done = False
+
+    def get_config_prop(self, name: str, default=None):
+        """
+        Get a specified config property.
+        Args:
+            name: name of the property
+            default: default value to return if the property is not defined.
+        Returns:
+        """
+        if not self.config:
+            return default
+        return self.config.get(name, default)
+
+    def start_run(self, fl_ctx: FLContext):
+        self.engine = fl_ctx.get_engine()
+        if not self.engine:
+            self.system_panic("no engine", fl_ctx)
+            return
+
+        runner = fl_ctx.get_prop(FLContextKey.RUNNER)
+        if not runner:
+            self.system_panic("no client runner", fl_ctx)
+            return
+
+        self.me = fl_ctx.get_identity_name()
+
+        self.persistor = self.engine.get_component(self.persistor_id)
+        if not isinstance(self.persistor, LearnablePersistor):
+            self.log_warning(
+                fl_ctx, f"Persistor {self.persistor_id} must be a Persistor instance, but got {type(self.persistor)}"
+            )
+            self.persistor = None
+
+        self.initialize(fl_ctx)
+
+    def initialize_controller(self, controller_id, fl_ctx):
+        controller = self.engine.get_component(controller_id)
+
+        comm = WFCommClient()
+        controller.set_communicator(comm)
+        controller.config = self.config
+        controller.initialize(fl_ctx)
+
+        return controller
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == EventType.START_RUN:
+            self.start_run(fl_ctx)
+
+        elif event_type == EventType.BEFORE_PULL_TASK:
+            # add my status to fl_ctx
+            if not self.workflow_id:
+                return
+
+            reports = fl_ctx.get_prop(Constant.STATUS_REPORTS)
+            if reports:
+                reports.pop(self.workflow_id, None)
+
+            if self.workflow_done:
+                return
+            report = self._get_status_report()
+            if not report:
+                self.log_debug(fl_ctx, "nothing to report this time")
+                return
+            self._add_status_report(report, fl_ctx)
+            self.last_status_report_time = report.timestamp
+
+        elif event_type in [EventType.ABORT_TASK, EventType.END_RUN]:
```

Comment:
Do we want `ABORT_TASK` here to shutdown the executor?
What if after `ABORT_TASK` this client executor receive a new task from server, what will happen?

Reply:
Right if `ABORT_TASK`, then the executor should be shutdown. If after `ABORT_TASK` the client executor receives a new task from the server, I can add a check at the beginning of `execute()` to see if `self.workflow_done`, which would be set in `finalize()`.

Reply:
ok, so after `ABORT_TASK`, this executor should be shutdown.

If any chances in the future where we could "re-start" a task, then we should create a new instance of client controller executor?

Reply:
I think since the ServerSideController and ClientControllerExecutor are paired, then we shouldn’t need to re-create an instance of the ClientControllerExecutor. A different ServerSideController could have a different corresponding ClientControllerExecutor however.

Reply:
Just hypothetical discussion so don't need to resolve in this PR.

For the same pair of ServerSideController and ClientControllerExecutor, if we can re-start a task in the future, we should think of how that is possible with these two classes.